### PR TITLE
MaxPartitionStrategy support file source based table.

### PR DESCRIPTION
**Pull request by cfmcgrady**
### _Why are the changes needed?_

Fix [comment](https://github.com/apache/incubator-kyuubi/pull/1086#discussion_r707103617)

### _How was this patch tested?_
- [x] Add some test cases that check the changes thoroughly including negative and positive cases if possible

- [ ] Add screenshots for manual tests if appropriate

- [x] [Run test](https://kyuubi.readthedocs.io/en/latest/develop_tools/testing.html#running-tests) locally before make a pull request

### CODE REVIEW

File: dev/kyuubi-extension-spark-3-1/src/main/scala/org/apache/kyuubi/sql/watchdog/MaxPartitionStrategy.scala [hunk 1]
```diff
@@ -0,0 +1,176 @@
...
+                  fsRelation @ HadoopFsRelation(
+                    fileIndex: InMemoryFileIndex,
+                    partitionSchema,
+                    _,
+                    _,
+                    _,
+                    _),
+                  _,
+                  _,
+                  _)) if fsRelation.partitionSchemaOption.isDefined =>
```
Comment (ulysses-you) [inline dev/kyuubi-extension-spark-3-1/src/main/scala/org/apache/kyuubi/sql/watchdog/MaxPartitionStrategy.scala:94]: actually, I really don't like such code format ... Is there a way to relax the new line style ? cc @pan3793 @zhouyifan279

File: dev/kyuubi-extension-spark-3-1/src/main/scala/org/apache/kyuubi/sql/watchdog/MaxPartitionStrategy.scala [hunk 2]
```diff
@@ -0,0 +1,176 @@
...
+                } else {
+                  Nil
+                }
+            }
+          case ScanOperation(
+                _,
+                filters,
+                relation @ LogicalRelation(
+                  fsRelation @ HadoopFsRelation(
+                    fileIndex: InMemoryFileIndex,
```
Comment (ulysses-you) [inline dev/kyuubi-extension-spark-3-1/src/main/scala/org/apache/kyuubi/sql/watchdog/MaxPartitionStrategy.scala:86]: does this `InMemoryFileIndex` only comes from `PruneFileSourcePartitions` who's relation has `partitionKeyFilters` ?

Comment (cfmcgrady) [reply dev/kyuubi-extension-spark-3-1/src/main/scala/org/apache/kyuubi/sql/watchdog/MaxPartitionStrategy.scala:86]: we need both to check a ScanOperation w/ or w/o partition key filters

- a ScanOperation with partition key filters

```sql
SELECT * FROM test where p = 1
```

partitionKeyFilters = ExpressionSet(isnotnull(p#45), (p#45 = 1))

only 1 partition will be scan

- a ScanOperation without partition key filters

```sql
SELECT * FROM test where
```
partitionKeyFilters = ExpressionSet()

as `partitionKeyFilters` is empty, `fileIndex.listFiles(partitionKeyFilters.toSeq, dataFilter)` will return all of partitions



So this strategy will be applied to all of the file-based partitioned tables.

File: dev/kyuubi-extension-spark-3-1/src/main/scala/org/apache/kyuubi/sql/watchdog/MaxPartitionStrategy.scala (hunk 1 above)
Comment (yaooqinn) [reply dev/kyuubi-extension-spark-3-1/src/main/scala/org/apache/kyuubi/sql/watchdog/MaxPartitionStrategy.scala:94]: this is not good..

File: dev/kyuubi-extension-spark-3-1/src/test/scala/org/apache/spark/sql/WatchDogSuite.scala [hunk 3]
```diff
@@ -31,38 +72,86 @@ class WatchDogSuite extends KyuubiSparkSQLExtensionTest {
...
+          .partitionBy("p")
+          .mode("overwrite")
+          .save(dir.getCanonicalPath)
+        spark.read.load(dir.getCanonicalPath).createOrReplaceTempView("test")
+        checkMaxPartition
+      }
+    }
+  }
 
+  test("watchdog with scan maxPartitions -- dynamic partition pruning") {
```
Comment (ulysses-you) [inline dev/kyuubi-extension-spark-3-1/src/test/scala/org/apache/spark/sql/WatchDogSuite.scala:133]: does this test necessary  ? seems it's same with `watchdog with scan maxPartitions -- data source`.

File: dev/kyuubi-extension-spark-3-1/src/main/scala/org/apache/kyuubi/sql/watchdog/MaxPartitionStrategy.scala [hunk 4]
```diff
@@ -0,0 +1,176 @@
...
+    new MaxPartitionExceedException(
+      s"""
+         |SQL job scan data source partition: $prunedPartitionSize
+         |exceed restrict of data source scan maxPartition $maxPartitionSize
+         |You should optimize your SQL logical according partition structure
+         |or shorten query scope such as p_date, detail as below:
+         |Table: ${tableMeta.map(_.qualifiedName).getOrElse("")}
+         |Owner: ${tableMeta.map(_.owner).getOrElse("")}
+         |RootPaths: ${fileIndex.mkString(",")}
+         |Partition Structure: ${partitionSchema.map(_.name).mkString(" -> ")}
```
Comment (ulysses-you) [inline dev/kyuubi-extension-spark-3-1/src/main/scala/org/apache/kyuubi/sql/watchdog/MaxPartitionStrategy.scala:173]: can you provide an example how this msg looks like ?

File: dev/kyuubi-extension-spark-3-1/src/main/scala/org/apache/kyuubi/sql/watchdog/MaxPartitionStrategy.scala [hunk 5]
```diff
@@ -0,0 +1,176 @@
...
+ * 1 Check if scan exceed maxPartition
+ * 2 Check if Using partitionFilter on partitioned table
+ * This Strategy Add Planner Strategy after LogicalOptimizer
+ */
+case class MaxPartitionStrategy(session: SparkSession)
+  extends Strategy
+  with SQLConfHelper
+  with PruneFileSourcePartitionHelper {
+  override def apply(plan: LogicalPlan): Seq[SparkPlan] = {
+    conf.getConf(KyuubiSQLConf.WATCHDOG_MAX_PARTITIONS) match {
```
Comment (ulysses-you) [inline dev/kyuubi-extension-spark-3-1/src/main/scala/org/apache/kyuubi/sql/watchdog/MaxPartitionStrategy.scala:43]: the code is hard readable. can we refactor it like:
```
val maxPartitions = conf.getConf(KyuubiSQLConf.WATCHDOG_MAX_PARTITIONS)
if (maxPartitions.isDefined) {
   checkRelationMaxPartitions(maxPartitions)
}

Nil
```

Comment (ulysses-you) [inline dev/kyuubi-extension-spark-3-1/src/main/scala/org/apache/kyuubi/sql/watchdog/MaxPartitionStrategy.scala:43]: the code is hard to read. can we refactor it like:
```scala
val maxPartitions = conf.getConf(KyuubiSQLConf.WATCHDOG_MAX_PARTITIONS)
if (maxPartitions.isDefined) {
   checkRelationMaxPartitions(maxPartitions)
}

Nil
```

Comment (ulysses-you) [reply dev/kyuubi-extension-spark-3-1/src/main/scala/org/apache/kyuubi/sql/watchdog/MaxPartitionStrategy.scala:43]: BTW, the code style can be formatted by `./dev/reformat` although it should be improved

Comment (cfmcgrady) [reply dev/kyuubi-extension-spark-3-1/src/main/scala/org/apache/kyuubi/sql/watchdog/MaxPartitionStrategy.scala:43]: Actually, it has been formatted already.

File: dev/kyuubi-extension-spark-3-1/src/test/scala/org/apache/spark/sql/WatchDogSuite.scala (hunk 3 above)
Comment (cfmcgrady) [reply dev/kyuubi-extension-spark-3-1/src/test/scala/org/apache/spark/sql/WatchDogSuite.scala]: removed.

File: dev/kyuubi-extension-spark-3-1/src/main/scala/org/apache/kyuubi/sql/watchdog/MaxPartitionStrategy.scala (hunk 4 above)
Comment (cfmcgrady) [reply dev/kyuubi-extension-spark-3-1/src/main/scala/org/apache/kyuubi/sql/watchdog/MaxPartitionStrategy.scala:171]: - textfile hive table w/ partition key filters
```
SQL job scan hive partition: 9
exceed restrict of hive scan maxPartition 5
You should optimize your SQL logical according partition structure
or shorten query scope such as p_date, detail as below:
Table: default.test
Owner: fchen
Partition Structure: p
```

- textfile hive table w/o partition key filters
```
Your SQL job scan a whole huge table without any partition filter,
You should optimize your SQL logical according partition structure
or shorten query scope such as p_date, detail as below:
Table: default.test
Owner: fchen
Partition Structure: p
```

- parquet hive table
```
SQL job scan data source partition: 10
exceed restrict of data source scan maxPartition 5
You should optimize your SQL logical according partition structure
or shorten query scope such as p_date, detail as below:
Table: default.test
Owner: fchen
RootPaths: file:/private/var/folders/b0/258422qj0n7bybfjb1gzvz3c0000gn/T/spark-c59b6bb7-2c66-4dd5-9774-4bf60850e995/org.apache.spark.sql.WatchDogSuite/warehouse/test
Partition Structure: p
```

- data source

```
SQL job scan data source partition: 6
exceed restrict of data source scan maxPartition 5
You should optimize your SQL logical according partition structure
or shorten query scope such as p_date, detail as below:
Table: 
Owner: 
RootPaths: file:/private/var/folders/b0/258422qj0n7bybfjb1gzvz3c0000gn/T/spark-7a07fd59-c94f-46cf-8bf6-bbeb2a494697
Partition Structure: p
```

File: dev/kyuubi-extension-spark-3-1/src/main/scala/org/apache/kyuubi/sql/watchdog/MaxPartitionStrategy.scala (hunk 1 above)
Comment (cfmcgrady) [reply dev/kyuubi-extension-spark-3-1/src/main/scala/org/apache/kyuubi/sql/watchdog/MaxPartitionStrategy.scala]: kindly ping @zhouyifan279 @pan3793

File: dev/kyuubi-extension-spark-3-1/src/main/scala/org/apache/kyuubi/sql/watchdog/MaxPartitionStrategy.scala (hunk 4 above)
Comment (ulysses-you) [reply dev/kyuubi-extension-spark-3-1/src/main/scala/org/apache/kyuubi/sql/watchdog/MaxPartitionStrategy.scala:178]: what happens if there are more one partitions ?

File: dev/kyuubi-extension-spark-3-1/src/main/scala/org/apache/kyuubi/sql/watchdog/MaxPartitionStrategy.scala [hunk 6]
```diff
@@ -0,0 +1,181 @@
...
+  extends Strategy
+  with SQLConfHelper
+  with PruneFileSourcePartitionHelper {
+  override def apply(plan: LogicalPlan): Seq[SparkPlan] = {
+    val maxScanPartitionsOpt = conf.getConf(KyuubiSQLConf.WATCHDOG_MAX_PARTITIONS)
+
+    if (maxScanPartitionsOpt.isDefined) {
+      val maxScanPartitions = maxScanPartitionsOpt.get
+
+      plan match {
```
Comment (ulysses-you) [inline dev/kyuubi-extension-spark-3-1/src/main/scala/org/apache/kyuubi/sql/watchdog/MaxPartitionStrategy.scala:48]: can we move this to a new method ?

File: dev/kyuubi-extension-spark-3-1/src/main/scala/org/apache/kyuubi/sql/watchdog/MaxPartitionStrategy.scala (hunk 4 above)
Comment (cfmcgrady) [reply dev/kyuubi-extension-spark-3-1/src/main/scala/org/apache/kyuubi/sql/watchdog/MaxPartitionStrategy.scala:182]:
```
SQL job scan data source partition: 9
exceed restrict of data source scan maxPartition 5
You should optimize your SQL logical according partition structure
or shorten query scope such as p_date, detail as below:
Table: default.test
Owner: fchen
RootPaths: file:/private/var/folders/b0/258422qj0n7bybfjb1gzvz3c0000gn/T/spark-eed43e68-d904-4aba-9414-7ea769fdf33b/org.apache.spark.sql.WatchDogSuite/warehouse/test/p=0,file:/private/var/folders/b0/258422qj0n7bybfjb1gzvz3c0000gn/T/spark-eed43e68-d904-4aba-9414-7ea769fdf33b/org.apache.spark.sql.WatchDogSuite/warehouse/test/p=2,file:/private/var/folders/b0/258422qj0n7bybfjb1gzvz3c0000gn/T/spark-eed43e68-d904-4aba-9414-7ea769fdf33b/org.apache.spark.sql.WatchDogSuite/warehouse/test/p=3,file:/private/var/folders/b0/258422qj0n7bybfjb1gzvz3c0000gn/T/spark-eed43e68-d904-4aba-9414-7ea769fdf33b/org.apache.spark.sql.WatchDogSuite/warehouse/test/p=4,file:/private/var/folders/b0/258422qj0n7bybfjb1gzvz3c0000gn/T/spark-eed43e68-d904-4aba-9414-7ea769fdf33b/org.apache.spark.sql.WatchDogSuite/warehouse/test/p=5... 4 more paths
Partition Structure: p
```

File: dev/kyuubi-extension-spark-3-1/src/main/scala/org/apache/kyuubi/sql/watchdog/MaxPartitionStrategy.scala (hunk 6 above)
Comment (cfmcgrady) [reply dev/kyuubi-extension-spark-3-1/src/main/scala/org/apache/kyuubi/sql/watchdog/MaxPartitionStrategy.scala]: Done.

File: dev/kyuubi-extension-spark-3-1/src/main/scala/org/apache/kyuubi/sql/watchdog/MaxPartitionStrategy.scala (hunk 4 above)
Comment (cfmcgrady) [reply dev/kyuubi-extension-spark-3-1/src/main/scala/org/apache/kyuubi/sql/watchdog/MaxPartitionStrategy.scala:182]:
```scala
    withTempDir { dir =>
      withTempView("test") {
        spark.range(10)
          .selectExpr("id", "explode(sequence(0, 10)) as v")
          .selectExpr("id", "id as p", "v as p2")
          .write
          .partitionBy("p", "p2")
          .mode("overwrite")
          .save(dir.getCanonicalPath)
        spark.read.load(dir.getCanonicalPath).createOrReplaceTempView("test")
        withSQLConf(KyuubiSQLConf.WATCHDOG_MAX_PARTITIONS.key -> "5") {
          sql("SELECT * FROM test where p=1").queryExecution.sparkPlan
        }
      }
    }
```

```
SQL job scan data source partition: 11
exceed restrict of data source scan maxPartition 5
You should optimize your SQL logical according partition structure
or shorten query scope such as p_date, detail as below:
Table: 
Owner: 
RootPaths: file:/private/var/folders/b0/258422qj0n7bybfjb1gzvz3c0000gn/T/spark-cbda132a-e5b0-4f66-849a-f566b730f7ec
Partition Structure: p -> p2
```

Comment (ulysses-you) [reply dev/kyuubi-extension-spark-3-1/src/main/scala/org/apache/kyuubi/sql/watchdog/MaxPartitionStrategy.scala:182]: `Partition Structure: p -> p2`
should it be ?
`Partition Structure: p, p2`

Comment (cfmcgrady) [reply dev/kyuubi-extension-spark-3-1/src/main/scala/org/apache/kyuubi/sql/watchdog/MaxPartitionStrategy.scala]: Yeah, updated.

**Review by ulysses-you — APPROVED** [review]

**Comment by ulysses-you** [comment]
thanks, merging to master/1.4

**Pull request closed without merging**